src/Data/Variant.hs

```diff
@@ -276,6 +279,36 @@ class CouldBeF xs x => CouldBe (xs :: [Type]) (x :: Type) where
 instance CouldBeF xs x => CouldBe xs x where
   throw = throwF . Identity
 
+type family FlattenConstraints (cs :: [Constraint]) = (c :: Constraint) | c -> cs where
+  FlattenConstraints  '[] = ()
+  FlattenConstraints (c ': cs) = (c, FlattenConstraints cs)
+
+type family TypeMap (f :: k -> l) (xs :: [k]) = (c :: [l]) where
```

Comment:
Maybe `ys` or `xs'` is a better name for `c` here, and I'd probably just call it `Map` now we're in a higher namespace :)